File: sdk/go/config_test.go
package sdk

import (
	"testing"
)

// Cat 02: Config validation tests (8 tests)

func TestValidDirectConfigPasses(t *testing.T) {
	cfg := SDKConfig{
		Chain: ChainConfig{
			ChainID:      "seocheon-test-1",
			RPCEndpoint:  "http://localhost:26657",
			GRPCEndpoint: "http://localhost:1317",
		},
		Signing: SigningConfig{
			Mode:     SigningModeDirect,
			Mnemonic: "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about",
		},
	}
	if err := cfg.Validate(); err != nil {
		t.Fatalf("expected valid config, got error: %v", err)
	}
}

func TestMissingChainIdFails(t *testing.T) {
	cfg := SDKConfig{
		Chain: ChainConfig{
			RPCEndpoint:  "http://localhost:26657",
			GRPCEndpoint: "http://localhost:1317",
		},
		Signing: SigningConfig{
			Mode:     SigningModeDirect,
			Mnemonic: "test mnemonic here",
		},
	}
	if err := cfg.Validate(); err == nil {
		t.Fatal("expected error for missing chain_id")
	}
}

func TestMissingRpcEndpointFails(t *testing.T) {
	cfg := SDKConfig{
		Chain: ChainConfig{
			ChainID:      "seocheon-test-1",
			GRPCEndpoint: "http://localhost:1317",
		},
		Signing: SigningConfig{
			Mode:     SigningModeDirect,
			Mnemonic: "test mnemonic here",
		},
	}
	if err := cfg.Validate(); err == nil {
		t.Fatal("expected error for missing rpc_endpoint")
	}
}

func TestMissingGrpcEndpointFails(t *testing.T) {
	cfg := SDKConfig{
		Chain: ChainConfig{
			ChainID:     "seocheon-test-1",
			RPCEndpoint: "http://localhost:26657",
		},
		Signing: SigningConfig{
			Mode:     SigningModeDirect,
			Mnemonic: "test mnemonic here",
		},
	}
	if err := cfg.Validate(); err == nil {
		t.Fatal("expected error for missing grpc_endpoint")
	}
}

func TestInvalidSigningModeFails(t *testing.T) {
	cfg := SDKConfig{
		Chain: ChainConfig{
			ChainID:      "seocheon-test-1",
			RPCEndpoint:  "http://localhost:26657",
			GRPCEndpoint: "http://localhost:1317",
		},
		Signing: SigningConfig{
			Mode: "invalid-mode",
		},
	}
	if err := cfg.Validate(); err == nil {
		t.Fatal("expected error for invalid signing mode")
	}
}

func TestDirectRequiresMnemonic(t *testing.T) {
	cfg := SDKConfig{
		Chain: ChainConfig{
			ChainID:      "seocheon-test-1",
			RPCEndpoint:  "http://localhost:26657",
			GRPCEndpoint: "http://localhost:1317",
		},
		Signing: SigningConfig{
			Mode: SigningModeDirect,
			// no mnemonic
		},
	}
	if err := cfg.Validate(); err == nil {
		t.Fatal("expected error when direct mode has no mnemonic")
	}
}

func TestVaultRequiresEndpointAndKey(t *testing.T) {
	cfg := SDKConfig{
		Chain: ChainConfig{
			ChainID:      "seocheon-test-1",
			RPCEndpoint:  "http://localhost:26657",
			GRPCEndpoint: "http://localhost:1317",
		},
		Signing: SigningConfig{
			Mode: SigningModeVault,
			// no vault_endpoint or key_name
		},
	}
	if err := cfg.Validate(); err == nil {
		t.Fatal("expected error when vault mode has no endpoint/key")
	}
}

func TestKeystoreRequiresPathAndPass(t *testing.T) {
	cfg := SDKConfig{
		Chain: ChainConfig{
			ChainID:      "seocheon-test-1",
			RPCEndpoint:  "http://localhost:26657",
			GRPCEndpoint: "http://localhost:1317",
		},
		Signing: SigningConfig{
			Mode: SigningModeKeystore,
			// no keystore_path or passphrase
		},
	}
	if err := cfg.Validate(); err == nil {
		t.Fatal("expected error when keystore mode has no path/passphrase")
	}
}

File: sdk/kotlin/src/test/kotlin/com/seocheon/sdk/ConfigTest.kt
package com.seocheon.sdk

import com.seocheon.sdk.errors.SdkError
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class ConfigTest {

    private fun validConfig(
        mode: SigningMode = SigningMode.DIRECT,
        mnemonic: String = "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about",
    ) = SDKConfig(
        chain = ChainConfig("test-chain", "http://localhost:26657", "http://localhost:9090"),
        signing = SigningConfig(mode = mode, mnemonic = mnemonic),
    )

    @Test
    fun `valid direct config passes validation`() {
        assertDoesNotThrow { validConfig().validate() }
    }

    @Test
    fun `empty chain_id fails validation`() {
        val config = SDKConfig(
            chain = ChainConfig("", "http://localhost:26657", "http://localhost:9090"),
            signing = SigningConfig(mode = SigningMode.DIRECT, mnemonic = "test mnemonic phrase"),
        )
        assertThrows(SdkError.InvalidConfig::class.java) { config.validate() }
    }

    @Test
    fun `empty rpc_endpoint fails validation`() {
        val config = SDKConfig(
            chain = ChainConfig("test-chain", "", "http://localhost:9090"),
            signing = SigningConfig(mode = SigningMode.DIRECT, mnemonic = "test"),
        )
        assertThrows(SdkError.InvalidConfig::class.java) { config.validate() }
    }

    @Test
    fun `empty grpc_endpoint fails validation`() {
        val config = SDKConfig(
            chain = ChainConfig("test-chain", "http://localhost:26657", ""),
            signing = SigningConfig(mode = SigningMode.DIRECT, mnemonic = "test"),
        )
        assertThrows(SdkError.InvalidConfig::class.java) { config.validate() }
    }

    @Test
    fun `vault mode requires endpoint and key_name`() {
        val config = SDKConfig(
            chain = ChainConfig("test-chain", "http://localhost:26657", "http://localhost:9090"),
            signing = SigningConfig(mode = SigningMode.VAULT),
        )
        assertThrows(SdkError.InvalidConfig::class.java) { config.validate() }
    }

    @Test
    fun `keystore mode requires path and passphrase`() {
        val config = SDKConfig(
            chain = ChainConfig("test-chain", "http://localhost:26657", "http://localhost:9090"),
            signing = SigningConfig(mode = SigningMode.KEYSTORE),
        )
        assertThrows(SdkError.InvalidConfig::class.java) { config.validate() }
    }

    @Test
    fun `direct mode requires mnemonic`() {
        val config = SDKConfig(
            chain = ChainConfig("test-chain", "http://localhost:26657", "http://localhost:9090"),
            signing = SigningConfig(mode = SigningMode.DIRECT),
        )
        assertThrows(SdkError.InvalidConfig::class.java) { config.validate() }
    }

    @Test
    fun `default config factory sets defaults`() {
        val config = SDKConfig.default(
            "test-chain",
            "http://localhost:26657",
            "http://localhost:9090",
            SigningConfig(mode = SigningMode.DIRECT, mnemonic = "test"),
        )
        assertEquals("sync", config.tx.broadcastMode)
        assertEquals(30000L, config.tx.confirmTimeoutMs)
        assertEquals(1000L, config.tx.confirmPollIntervalMs)
    }
}
